### PR TITLE
feat(snippet_rules): add v4 to v5 migration

### PR DESCRIPTION
**Pull request by musa-cf**
Screenshot 2026-01-07 at 6 47 09 AM

Integration test output:
```
╰─❯ TEST_RESOURCE=snippet_rules go test -v -run TestSingleResource ./integration/v4_to_v5/

=== RUN   TestSingleResource
=== RUN   TestSingleResource/snippet_rules
--- PASS: TestSingleResource (0.84s)
    --- PASS: TestSingleResource/snippet_rules (0.84s)
PASS
ok  	github.com/cloudflare/tf-migrate/integration/v4_to_v5	1.410s
```

### CODE REVIEW

File: integration/v4_to_v5/testdata/snippet_rules/input/terraform.tfstate.backup
Comment (ssicard) [inline integration/v4_to_v5/testdata/snippet_rules/input/terraform.tfstate.backup:1]: Please delete this file.

**Review by ssicard — CHANGES_REQUESTED** [review]

File: integration/v4_to_v5/testdata/snippet_rules/input/snippet_rules_e2e.tf [hunk 1]
```diff
@@ -0,0 +1,293 @@
+# Comprehensive Integration Tests for snippet_rules Migration
+# This file tests all Terraform patterns and edge cases
+
+# Test 1: Basic resource with single rule
+resource "cloudflare_snippet_rules" "basic" {
```
Comment (SirCortly) [inline integration/v4_to_v5/testdata/snippet_rules/input/snippet_rules_e2e.tf:5]: I know this is accomplishing pretty much the same thing but I added a pattern where you can suffix a file such as `snippet_rules_e2e.tf` and it will only pick up on that file for the e2e tests, allowing the non suffixed .tf file to be comprehensive integration tests.

File: integration/v4_to_v5/testdata/snippet_rules/input/terraform.tfstate.backup
Comment (musa-cf) [reply integration/v4_to_v5/testdata/snippet_rules/input/terraform.tfstate.backup:1]: Done!

File: integration/v4_to_v5/testdata/snippet_rules/input/snippet_rules_e2e.tf (hunk 1 above)
Comment (musa-cf) [reply integration/v4_to_v5/testdata/snippet_rules/input/snippet_rules_e2e.tf:5]: Done!

**Review by tamas-jozsa — APPROVED** [review]

**Review by SirCortly — APPROVED** [review]

**Comment by vaishakdinesh** [comment]
```
~/cf-repos/github/tf-migrate snippet-rules ⇣2⇡88 !4 ❯ TF_MIG_TEST=true ./scripts/run-e2e-tests.sh --apply-exemptions --provider /Users/vaishak/cf-repos/github/terraform-provider-cloudflare --resources snippet
Building binaries...
go build -v -o bin/tf-migrate ./cmd/tf-migrate
go build -v -o bin/e2e-runner ./cmd/e2e-runner
github.com/cloudflare/tf-migrate/cmd/e2e-runner


========================================
State Migration Methods for Specified Resources
========================================

Resources using tf-migrate TransformState:
  snippet

Targeting specific resources: snippet
Target arguments: -target=module.snippet


========================================
E2E Migration Test
========================================

Step 0: Initializing test resources
Running tests with:
  User:       vaishak@cloudflare.com
  Account ID: 8511639f53e8d7871003f714afd1dba4
  Zone ID:    b9c9ae341a26b5b692441bb6573015b9
  Domain:     vaishak.terraform.cfapi.net
  Provider:   Local (/Users/vaishak/cf-repos/github/terraform-provider-cloudflare)

Running init script...

========================================
Syncing Test Resources
========================================

Filtering to specific resources: snippet
Syncing resource files from testdata...
  ✓ snippet/snippet.tf
  ✓ snippet/versions.tf

  Total: 2 files synced

Configuring terraform variables...


✓ Saved configuration
    Account ID: 8511639f53e8d7871003f714afd1dba4
    Zone ID: b9c9ae341a26b5b692441bb6573015b9
    Domain: vaishak.terraform.cfapi.net
    File: v4/terraform.tfvars

Scanning for import annotations...

Updating main.tf with module references...
  ↻ Updated main.tf with 1 module references


========================================
✓ Sync Complete!
========================================

Summary:
  - Terraform v4 configs: /Users/vaishak/cf-repos/github/tf-migrate/e2e/tf/v4
  - Modules: 1
  - Files synced: 2

Configuring remote backend...
✓ Backend configured

  ✓ Provider installation preserved
  ✓ Backend already configured

Next steps:
  cd tf/v4 && terraform apply

Note: Configuration is automatically loaded from terraform.tfvars
      State is managed remotely in R2

✓ Test resources initialized


========================================
Setting up local provider
========================================

Using provider from: /Users/vaishak/cf-repos/github/terraform-provider-cloudflare
Building provider...
  Building in: /Users/vaishak/cf-repos/github/terraform-provider-cloudflare
  Output: /Users/vaishak/cf-repos/github/terraform-provider-cloudflare/terraform-provider-cloudflare
✓ Provider built successfully: /Users/vaishak/cf-repos/github/terraform-provider-cloudflare/terraform-provider-cloudflare
✓ Created dev overrides config: /Users/vaishak/cf-repos/github/tf-migrate/.terraformrc-tf-migrate
✓ Local provider will be used for v5 testing

Note: v4 tests will use the registry provider (v4.x)
      v5 tests will use the local provider with dev overrides

Step 1: Testing v4 configurations
Running terraform init in v4/...
Found local state file, backing up and using remote state...
✓ Terraform init successful (remote state loaded from R2)
Running terraform plan in v4/...
✓ Terraform plan shows no changes

Running terraform apply in v4/...
✓ Terraform apply successful
  Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
Syncing state from remote...
✓ Local state file synced from R2
Capturing v4 state...
✓ Saved v4 state to tmp/v4-state.json


Step 2: Running migration
Running ./scripts/migrate...
Building tf-migrate binary...
✓ Binary built successfully

========================================
Running v4 to v5 Migration
========================================

Preparing output directory...
  ✓ Preserved v5 provider installation (.terraform/)
  ✓ Preserved v5 dependency lock file (.terraform.lock.hcl)
Copying only targeted resources: snippet
    ✓ Copied root file: provider.tf
    ✓ Copied root file: terraform.tfvars
    ✓ Copied root file: terraform.tfstate

    ✓ Copied module: snippet
Creating filtered main.tf...
✓ Copied targeted resources to migrated-v4_to_v5/
✓ Updated provider.tf to use ~> 5.0 and removed backend config
Filtering state file to only include targeted resources...
✓ Filtered state to 19 resources from targeted modules

Migrating configuration files...
Cloudflare Terraform Provider Migration Tool
============================================

Configuration directory: /Users/vaishak/cf-repos/github/tf-migrate/e2e/migrated-v4_to_v5
Output directory: in-place
✓ Using Cloudflare API credentials (API key + email)

Found 4 configuration files to migrate
[1/4] Processing main.tf... ✓
[2/4] Processing provider.tf... ✓
[3/4] Processing snippet.tf... ✓
[4/4] Processing versions.tf... ✓
2026-02-19T08:55:31.442-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*zone_setting.V4ToV5Migrator"
2026-02-19T08:55:31.443-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*zero_trust_split_tunnel.V4ToV5Migrator"
2026-02-19T08:55:31.443-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*rulesets.V4ToV5Migrator"
2026-02-19T08:55:31.443-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*queue.V4ToV5Migrator"
2026-02-19T08:55:31.443-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*load_balancer_pools.V4ToV5Migrator"
2026-02-19T08:55:31.443-0800 [WARN]  tf-migrate: Migrator does not implement ResourceRenamer interface - cross-file references may not be updated: migrator="*zero_trust_gateway_certificate.V4ToV5Migrator"

Applying cross-file reference updates (29 updates across 4 files)...
✓ Updated cross-file references (29 updates applied)

Processing state file: terraform.tfstate... ✓
✓ Migration complete (config transformed, state will be upgraded by provider)


========================================
✓ Migration Complete!
========================================

Results:
  Input (v4):  /Users/vaishak/cf-repos/github/tf-migrate/e2e/tf/v4
  Output (v5): /Users/vaishak/cf-repos/github/tf-migrate/e2e/migrated-v4_to_v5

Next steps:
  cd /Users/vaishak/cf-repos/github/tf-migrate/e2e/migrated-v4_to_v5
  terraform init
  terraform plan

✓ Migration successful

Step 3: Testing v5 configurations
Running terraform init in migrated-v4_to_v5/...
Cleaning v5 .terraform directory for fresh init...
Removing .terraform.lock.hcl to allow dev_overrides...
✓ Terraform init successful
Running terraform plan in v5/...
✓ Terraform plan shows no changes (expected)
Running terraform apply in v5/...
✓ Terraform apply successful
Capturing v5 state...
✓ Saved v5 state to tmp/v5-state.json

Step 4: Verifying stable state (v5 plan after apply)
Running terraform plan again to check for ongoing drift...
✓ No ongoing drift detected - migration achieved stable state!



========================================
✓ E2E Test Complete!
========================================

Summary:

  Step 1: v4 terraform apply
    Status: ✓ SUCCESS

  Step 2: Migration (v4 → v5)
    Status: ✓ SUCCESS

  Step 3: v5 plan (before apply)
    Status: ✓ No changes needed

  Step 4: v5 terraform apply
    Status: ✓ SUCCESS

  Step 5: v5 plan (after apply)
    Status: ✓ SUCCESS - Stable state achieved
    Result: No changes detected

Logs saved to:
  - /Users/vaishak/cf-repos/github/tf-migrate/e2e/tmp
  ```

File: integration/v4_to_v5/testdata/snippet_rules/expected/snippet_rules_e2e.tf
Comment (ssicard) [inline integration/v4_to_v5/testdata/snippet_rules/expected/snippet_rules_e2e.tf:1]: nit: pretty sure we only need e2e files in the input/ dir

**Review by ssicard — APPROVED** [review]

File: integration/v4_to_v5/testdata/snippet_rules/expected/snippet_rules_e2e.tf
Comment (vaishakdinesh) [reply integration/v4_to_v5/testdata/snippet_rules/expected/snippet_rules_e2e.tf:1]: Deleted!
